include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -0,0 +1,1842 @@
+                                     k * n_child_dofs_1d * n_child_dofs_1d +
+                                     j * n_child_dofs_1d + i + shift;
+    }
+
+    template <int dim>
+    std::vector<std::vector<unsigned int>>
+    get_child_offsets(const unsigned int n_dofs_coarse_cell,
+                      const unsigned int fe_shift_1d,
+                      const unsigned int fe_degree)
+    {
+      std::vector<std::vector<unsigned int>> cell_local_chilren_indices(
+        GeometryInfo<dim>::max_children_per_cell,
+        std::vector<unsigned int>(n_dofs_coarse_cell));
+      {
+        for (unsigned int c = 0; c < GeometryInfo<dim>::max_children_per_cell;
+             c++)
+          get_child_offset<dim>(c,
+                                fe_shift_1d,
+                                fe_degree,
+                                cell_local_chilren_indices[c]);
+      }
+      return cell_local_chilren_indices;
+    }
+
+    template <int dim>
+    types::coarse_cell_id
+    convert_cell_id_binary_type_to_level_coarse_cell_id(
+      const typename CellId::binary_type &binary_representation)
+    {
+      // exploiting the structure of CellId::binary_type
+      // see also the documentation of CellId
+
+      // actual coarse-grid id
+      const unsigned int coarse_cell_id  = binary_representation[0];
+      const unsigned int n_child_indices = binary_representation[1] >> 2;
+
+      const unsigned int children_per_value =
+        sizeof(CellId::binary_type::value_type) * 8 / dim;
+      unsigned int child_level  = 0;
+      unsigned int binary_entry = 2;
+
+      // path to the get to the cell
+      std::vector<unsigned int> cell_indices;
+      while (child_level < n_child_indices)
+        {
+          Assert(binary_entry < binary_representation.size(),
+                 ExcInternalError());
+
+          for (unsigned int j = 0; j < children_per_value; ++j)
+            {
+              unsigned int cell_index =
+                (((binary_representation[binary_entry] >> (j * dim))) &
+                 (GeometryInfo<dim>::max_children_per_cell - 1));
+              cell_indices.push_back(cell_index);
+              ++child_level;
+              if (child_level == n_child_indices)
+                break;
+            }
+          ++binary_entry;
+        }
+
+      // compute new coarse-grid id: c_{i+1} = c_{i}*2^dim + q;
+      types::coarse_cell_id level_coarse_cell_id = coarse_cell_id;
+      for (auto i : cell_indices)
+        level_coarse_cell_id =
+          level_coarse_cell_id * GeometryInfo<dim>::max_children_per_cell + i;
+
+      return level_coarse_cell_id;
+    }
+
+  } // namespace
+
+  class FineDoFHandlerViewCell
+  {
+  public:
+    FineDoFHandlerViewCell(
+      const std::function<bool()> &has_children_,
+      const std::function<void(std::vector<types::global_dof_index> &)>
+        &get_dof_indices_)
+      : has_children_(has_children_)
+      , get_dof_indices_(get_dof_indices_)
```

Comment:
Is there a reason to use the trailing underscores in the variable names? We otherwise don't use that convention in deal.II.